scripts/eval.py

```diff
@@ -0,0 +1,286 @@
+                # This prevents any parameters from being initialized twice
+                def dummy_init_fn(module: torch.nn.Module) -> None:
+                    module.to_empty(device=get_default_device())
+
+                param_init_fn = dummy_init_fn
+            else:
+                param_init_fn = None
+
+            # Set up device mesh for hybrid sharding in order to specify which nodes are assoicated to a given model replica
+            device_mesh = None
+            hybrid_sharding_fsdp_kwargs = {}
+            if cfg.fsdp.sharding_strategy in (ShardingStrategy.HYBRID_SHARD, ShardingStrategy._HYBRID_SHARD_ZERO2):
+                if version.parse(torch.__version__) < version.parse("2.2.0"):
+                    # Device mesh was not added to PyTorch until v2.2.0
+                    raise OLMoConfigurationError(
+                        "OLMo training does not correctly support hybrid sharding before torch 2.2.0"
+                    )
+
+                from torch.distributed.device_mesh import init_device_mesh
+
+                num_model_replicas = cfg.fsdp.hybrid_sharding_num_model_replicas or (
+                    get_world_size() // get_local_world_size()
+                )
+
+                if num_model_replicas <= 0:
+                    raise OLMoConfigurationError("fsdp.hybrid_sharding_num_model_replicas must be a positive integer")
+
+                if get_world_size() % num_model_replicas != 0:
+                    raise OLMoConfigurationError("fsdp.hybrid_sharding_num_model_replicas must divide world size")
+
+                device_mesh = init_device_mesh("cuda", (num_model_replicas, get_world_size() // num_model_replicas))
+                hybrid_sharding_fsdp_kwargs["device_mesh"] = device_mesh
+
+            dist_model = FSDP(
+                olmo_model,
+                sharding_strategy=cfg.fsdp.sharding_strategy,
+                mixed_precision=cfg.fsdp_precision,
+                auto_wrap_policy=wrap_policy,
+                use_orig_params=cfg.fsdp.use_orig_params,  # needed for compile and some of our optimizer/parameter metrics
+                limit_all_gathers=True,
+                device_id=get_local_rank(),
+                param_init_fn=param_init_fn,
+                **hybrid_sharding_fsdp_kwargs,
+            )
+        elif cfg.distributed_strategy is None:
+            raise NotImplementedError("Single accelerator training not implemented yet!")
+
+        # when param_init_fn is None, FSDP will call reset_parameters() automatically
+        if param_init_fn is not None or cfg.distributed_strategy == DistributedStrategy.ddp:
+            olmo_model.reset_parameters()
+
+        log.info(f"Peak GPU Memory (MB) after {cfg.distributed_strategy}: {int(peak_gpu_memory() or 0)}")
+        log.info("Model:")
+        log.info(dist_model)
+
+        # Construct optimizer and learning rate scheduler.
+        optim = build_optimizer(cfg, dist_model)
+        scheduler = build_scheduler(cfg)
```

Comment:
We don't need optimizers and schedulers if we're just evaluating.

Reply:
So you're creating these only so that you can produce a `Trainer` object?

How hard is it to pull the stuff you need out of the `Trainer` object, so we don't have to do so many things we don't need? It makes me particularly uncomfortable that you're creating a trainer with a `None` data loader, which isn't supposed to work. It just happens to work.

Reply:
The Trainer class has too many precious helper functions and it's kinda dumb to unroll them all. I do wanna keep at least a dummy Trainer object. Let me see if I can create it w/o the optim/scheduler/etc stuff.

Reply:
I think you might find that you don't need most of that stuff when you're doing inference only.